examples/megssm/mne_util.py

```diff
@@ -0,0 +1,290 @@
+            # these labels store both hemispheres so subtract the rh offset
+            # from that part of the vertex array
+            lh_label_vert_idx = self.fwd_src_roi[:self.n_lhverts, li].nonzero()[0]
+            rh_label_vert_idx = self.fwd_src_roi[self.n_lhverts:, li].nonzero()[0]
+            rh_label_vert_idx[self.n_lhverts:] -= self.offsets['rh']
+            return [lh_label_vert_idx, rh_label_vert_idx]
+
+    def get_label_verts(self, label, src):
+        # if you're thinking of using this to plot, why not just use
+        # brain.add_label from pysurfer?
+        if isinstance(label, mne.Label):
+            hi = 0 if label.hemi == 'lh' else 1
+            label_vert_idx = self.get_label_vinds(label)
+            varray = src[hi]['vertno'][label_vert_idx]
+        elif isinstance(label, mne.BiHemiLabel):
+            lh_label_vert_idx, rh_label_vert_idx = self.get_label_vinds(label)
+            varray = [src[0]['vertno'][lh_label_vert_idx],
+                      src[1]['vertno'][rh_label_vert_idx]]
+        return varray
+
+    def get_hemi_idx(self, label):
+        if isinstance(label, mne.Label):
+            return 0 if label.hemi == 'lh' else 1
+        elif isinstance(label, mne.BiHemiLabel):
+            hemis = [None] * 2
+            for i, lab in enumerate([label.lh, label.rh]):
+                hemis[i] = 0 if lab.hemi == 'lh' else 1
+            return hemis
+
+def apply_projs(epochs, fwd, cov):
+    """ apply projection operators to fwd and cov """
+    proj, _ = mne.io.proj.setup_proj(epochs.info, activate=False)
+    fwd_src_sn = fwd['sol']['data']
+    fwd['sol']['data'] = np.dot(proj, fwd_src_sn)
+
+    roi_cov = cov.data
+    if not np.allclose(np.dot(proj, roi_cov), roi_cov):
+        roi_cov = np.dot(proj, np.dot(roi_cov, proj.T))
+        cov.data = roi_cov
+
+    return fwd, cov
+
+
+def _scale_sensor_data(epochs, fwd, cov, roi_to_src, eeg_scale=1., mag_scale=1.,
+    grad_scale=1.):
+    """ apply per-channel-type scaling to epochs, forward, and covariance """
+    # # from util import Carray ##skip import just pasted; util also from MEGLDS repo
+    # Carray64 = lambda X: np.require(X, dtype=np.float64, requirements='C')
+    # Carray32 = lambda X: np.require(X, dtype=np.float32, requirements='C')
+    # Carray = Carray64
+
+    # get indices for each channel type
+    ch_names = cov['names'] # same as self.fwd['info']['ch_names']
+    sel_eeg = pick_types(fwd['info'], meg=False, eeg=True, ref_meg=False)
+    sel_mag = pick_types(fwd['info'], meg='mag', eeg=False, ref_meg=False)
+    sel_grad = pick_types(fwd['info'], meg='grad', eeg=False, ref_meg=False)
+    idx_eeg = [ch_names.index(ch_names[c]) for c in sel_eeg]
+    idx_mag = [ch_names.index(ch_names[c]) for c in sel_mag]
+    idx_grad = [ch_names.index(ch_names[c]) for c in sel_grad]
+
+    # retrieve forward and sensor covariance
+    fwd_src_snsr = fwd['sol']['data'].copy()
+    snsr_cov = cov.data.copy()
```

Comment:
... probably here you do
```
epochs = epochs.copy().pick('data', exclude='bads')
snsr_cov = pick_channels_cov(cov, epochs.ch_names, ordered=True).data
fwd = convert_forward_solution(fwd, force_fixed=True)
fwd_src_snsr = pick_channels_forward(fwd, epochs.ch_names, ordered=True)['sol']['data']
data = epochs.get_data()
info = epochs.info
del cov, fwd, epochs
```
now you are guaranteed that `data`, `fwd_src_snsr`, and `snsr_cov `all have the same set of channels in the same order. Then to rescale them all you can just do:
```
rescale_cov = make_ad_hoc_cov(epochs.info)
scaler = compute_whitener(info, rescale_cov)
del rescale_cov
fwd_src_snsr = scaler @ fwd_src_snsr
snsr_cov = scaler @ snsr_cov @ scaler.T
data = scaler @ epochs.get_data()  # @ nicely knows to operate over the last two dims of (epochs, chs, time)
```
or something similar. Basically you let the MNE `pick_channels_*` make sure all channels are present and ordered properly, then make use of MNE functions to do the rescaling.

Reply:
Previously you said to play around with these functions to see if it would produce the same outputs as the `scale_sensor_data` function. The original function was using a scale of 1 for each of the channel types so nothing was changed. This scaling does alter the data, primarily in a significant increase in the number of principal components (PCA run immediately after scaling function). Also, the output of the model is much noisier using `@ scaler`, I think as expected with the significant increase in PCs. Let's talk to make sure this is working as expected.

Reply:
Output using mne functions for scaling seems much noisier.
![image](https://user-images.githubusercontent.com/39603454/180334645-499c773d-20aa-4de5-ac65-3628f9eac992.png)

Reply:
Compared to no scaling (or previous function with scale of 1 for each channel type).
![image](https://user-images.githubusercontent.com/39603454/180334746-1d6af98f-1413-4254-bbf6-713a61412db4.png)

Reply:
> The original function was using a scale of 1 for each of the channel types so nothing was changed.

This is the equivalent of using `mne.make_ad_hoc_cov(..., std=1.)`. So if you want this mode you can get it easily with a very small change. Can you verify this looks the same as the old code?

> This scaling does alter the data, primarily in a significant increase in the number of principal components (PCA run immediately after scaling function).

This is to be expected. If you do no scaling and have MEG *and* EEG data, your EEG data will be ~6 orders of magnitude larger or so, so you will end up only looking at EEG data.

> Also, the output of the model is much noisier using @ scaler, I think as expected with the significant increase in PCs. Let's talk to make sure this is working as expected.

What's not clear to me: is it also noisier (and noisier in the same way) when using the *old* code path but *enabling* scaling?

In other words: is this a problem with the new way of scaling, or is this a problem that has always existed with scaling the data, regardless of scaling method?

We can chat about this today